src/ethernet/RsRtspClient.cpp

```diff
@@ -0,0 +1,613 @@
+  }
+  m_commandDone = false;
+  
+  if (m_lastReturnValue.exit_code!=RsRtspReturnCode::OK)
+  {
+    throw std::runtime_error(format_error_msg(__FUNCTION__, m_lastReturnValue));
+  }
+
+  if (this->m_supportedProfiles.size()==0)
+  {
+    RsRtspReturnValue err = 
+      {RsRtspReturnCode::ERROR_GENERAL, std::string("failed to get streams from network device at url: " + std::string(this->url()))};
+    throw std::runtime_error(format_error_msg(__FUNCTION__, err));
+  }
+
+  return this->m_supportedProfiles;
+}
+
+int RsRTSPClient::addStream(rs2_video_stream t_stream, rtp_callback *t_callbackObj)
+{
+  long long int uniqueKey = getStreamProfileUniqueKey(t_stream);
+  RsMediaSubsession *subsession = this->m_subsessionMap.find(uniqueKey)->second;
+  if(subsession == nullptr)
+  {
+    RsRtspReturnValue err = 
+      {RsRtspReturnCode::ERROR_WRONG_FLOW,"requested stream was not found"};
+    throw std::runtime_error(format_error_msg(__FUNCTION__, err));
+  }
+  
+    if (!subsession->initiate())
+    {
+      this->envir() << "Failed to initiate the subsession \n";
+        RsRtspReturnValue err = {RsRtspReturnCode::ERROR_WRONG_FLOW,"Failed to initiate the subsession"};
+      throw std::runtime_error(format_error_msg(__FUNCTION__,err));
+    }
+    else
+    {
+      // Continue setting up this subsession, by sending a RTSP "SETUP" command:
+      unsigned res = this->sendSetupCommand(*subsession, this->continueAfterSETUP, False, REQUEST_STREAMING_OVER_TCP);
+      // wait for continueAfterSETUP to finish
+      std::unique_lock<std::mutex> lck(m_commandMtx);
+      m_cv.wait_for(lck, std::chrono::seconds(RTSP_CLIENT_COMMANDS_TIMEOUT_SEC), [this] { return m_commandDone; });
+      // for the next command
+      if(!m_commandDone)
+      {
+      RsRtspReturnValue err = 
+        {RsRtspReturnCode::ERROR_TIME_OUT,"client time out"};
+      throw std::runtime_error(format_error_msg(__FUNCTION__, err));
+      }
+      m_commandDone = false;
+
+      if(m_lastReturnValue.exit_code!=RsRtspReturnCode::OK)
+      {
+        throw std::runtime_error(format_error_msg(__FUNCTION__,m_lastReturnValue));
+      }
+      else
+      {
+        subsession->sink = RsSink::createNew(this->envir(), *subsession, t_stream, m_memPool, this->url());
+        // perhaps use your own custom "MediaSink" subclass instead
+        if (subsession->sink == NULL)
+        {
+          this->envir() << "Failed to create a data sink for the subsession: " << this->envir().getResultMsg() << "\n";
+          RsRtspReturnValue err = 
+            {(RsRtspReturnCode)envir().getErrno(),std::string("Failed to create a data sink for the subsession: " + std::string(envir().getResultMsg()))};
+          throw std::runtime_error(format_error_msg(__FUNCTION__,err));
+          // TODO: define error
+        }
+
+        subsession->miscPtr = this; // a hack to let subsession handler functions get the "RTSPClient" from the subsession
```

Comment:
This is not safe. 
Should be replaced with creating and passing `weak_ptr` to subsession